src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,631 @@
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  SVDTransformer
+// |
+// ----------------------------------------------------------------------
+template <typename InputT, typename TransformedT>
+SVDTransformer<InputT, TransformedT>::SVDTransformer(TransformedT singularvalues, TransformedT singularvectors) :
+    //no validation here because same variable has been validated in annotation class
+    _singularvalues(std::move(singularvalues)),
+    _singularvectors(std::move(singularvectors)) {
+}
+
+//TODO:
+// template <typename InputT, typename TransformedT>
+// SVDTransformer<InputT, TransformedT>::SVDTransformer(Archive &ar) :
+//     SVDTransformer(
+//         [&ar](void) {
+//             // Version
+//             std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+//             std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+//             if(majorVersion != 1 || minorVersion != 0)
+//                 throw std::runtime_error("Unsupported archive version");
+
+//             // Data:TODO
+//             TransformedT                   singularvalues(Traits<TransformedT>::deserialize(ar));
+//             TransformedT                   singularvectors(Traits<TransformedT>::deserialize(ar));
+
+//             return SVDTransformer<InputT, TransformedT>(std::move(singularvalues), std::move(singularvectors));
+//         }()
+//     ) {
+// }
+
+template <typename InputT, typename TransformedT>
+bool SVDTransformer<InputT, TransformedT>::operator==(SVDTransformer const &other) const {
+    if ((this->_singularvalues - other._singularvectors).norm() > 0.000001f || (this->_singularvalues - other._singularvectors).norm() > 0.000001f)
+        return false;
+
+    return true;
+}
+
+template <typename InputT, typename TransformedT>
+void SVDTransformer<InputT, TransformedT>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data TODO
+    // Traits<decltype(_singularvalues)>::serialize(ar, _singularvalues);
+    // Traits<decltype(_singularvectors)>::serialize(ar, _singularvectors);
+}
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+template <typename InputT, typename TransformedT>
+void SVDTransformer<InputT, TransformedT>::execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) /*override*/ {
+    if (input.rows() == 0)
+        throw std::invalid_argument("Input matrix rows() invalid");
+    
+    if (input.cols() != _singularvectors.rows()) 
+        throw std::invalid_argument("Input matrix cols() invalid");
+
+    callback(std::move(input * _singularvectors));
```

Comment:
`input * _singlevectors` will create a temporary (which isn't assigned to a variable), so `move` isn't necessary here.